elastic/kibana
# [Upgrade Assistant] Add support for API keys when reindexing

Fixes https://github.com/elastic/kibana/issues/72014

Upon receiving a request to reindex, Upgrade Assistant stores the requester's credentials in memory and uses those credentials when updating saved objects and issuing requests for other reindexing-related operations. In a token-based authentication mechanism, these tokens have an expiration time that could expire before reindexing work completes causing the task to stall. 

As a partial fix, this PR implements granting a temporary API key that is stored in memory that is then used for issuing requests as advised via https://github.com/elastic/kibana/issues/72014#issuecomment-892394832. Once the reindex operation is completed, the API key is invalidated. Note: this implementation is only possible if API keys are enabled. If API keys are *not* enabled, we will have to continue using the current approach.

**How to test**

Unzip [6.8.16-data.zip](https://github.com/elastic/kibana/files/7007813/6.8.16-data.zip). This contains indices created on 6.8 to trigger the reindex action in UA.

1. **Test reindexing with API keys enabled.** To enable API keys, start ES as follows:

      ```
      yarn es snapshot -E xpack.security.authc.api_key.enabled=true -E path.data=/path/to/6.8.16-data
      ```

     - Navigate to UA --> Elasticsearch deprecations
     - Select an index that needs to be reindexed, and start the reindexing process
     - Once started, you can verify an API key was created by navigating to Security --> API keys
     - Verify the reindex process completed successfully 
     - Once the reindex process is complete, the API key should no longer appear under Security --> API keys

2. **Test reindexing with API keys disabled.** Restart ES:

      ```
      yarn es snapshot -E path.data=/path/to/6.8.16-data
      ```

     - Navigate to UA --> Elasticsearch deprecations
     - Select an index that needs to be reindexed, and start the reindexing process
     - Verify the reindex process completed successfully
